src/Tools/ExternalAccess/Razor/RazorDynamicFileInfo.cs

```diff
@@ -41,5 +41,15 @@ public RazorDynamicFileInfo(string filePath, SourceCodeKind sourceCodeKind, Text
         /// return <see cref="IRazorDocumentServiceProvider"/> for the contents it provides
         /// </summary>
         public IRazorDocumentServiceProvider DocumentServiceProvider { get; }
+
+        /// <summary>
+        /// Constructs a new <see cref="DocumentInfo"/> from an existing <see cref="DocumentInfo"/> but with updated
+        /// text loader and document service provider coming from this instance.
+        /// </summary>
+        public DocumentInfo ToUpdatedDocumentInfo(DocumentInfo existingDocumentInfo)
```

Comment:
Who was creating the original DocumentInfo? I'm surprised they couldn't have just created this directly?

Reply:
Razor creates it on Mac, but in VS its created by us
https://github.com/dotnet/razor-tooling/pull/7013/files#diff-586009e069eea2515066ca61a6c43b9722c82c7454b5159cc751fc87975718bdL99

Reply:
From the public API, there is no way to ever set or change the value of `DocumentServicesProvider`. On VS Win this isn't a problem, as Roslyn does the work in the VS layer. On VS Mac, even though it has IVT, it can't do the same work, for other reasons, so Razor has to do it. Hence there are three possible solutions to this I considered:

1. Update the public API of DocumentInfo. `DocumentServicesProvider` is the odd standout that can't be set in the constructor, and doesn't have a `With` method, so seems straight forward. 
2. Update VS Mac to do the same thing as the Roslyn VS layer is doing, and have it use its IVT to call the constructor
3. Poke the constructor out of the EA hole in a fashion exactly as you see before you

The reasons I ruled each idea out:

1. In the Long Dark Backlog of Time, eventually the whole notion of `DocumentPropertiesService` is desired to go away, and with it `DocumentServicesProvider` might make less sense etc. so whilst "allow updating `DocumentServicesProvider`" might be an easy thing to take to API review, in future it will be a _lot_ easier to take "Remove `DocumentServicesProvider`, which no consumer could ever observe as being non-null anyway" than "Remove `DocumentServicesProvider` which any consumer could use for arbitrary purposes".
2. This is a painful job. It requires an API change in VS Mac (their `GetDynamicDocumentInfo` method returns a `DocumentInfo`, when it needs to return a `DynamicDocumentInfo`) which would need to be done (binary breaking change for all consumers), so that a package could be produced with the new interface, so Razor can consume, then insert back into VS Mac. While this is underway everthing is broken. There is no such thing as a dual insertion, because the interface is defined in the thing being inserted to. So we're left with doing COM tricks, and we have an `IDynamicDocumentInfoProvider2` and the world is a horrible place. Plus it keeps VS Mac relying on IVT, which in the Even Longer Even Darker Backlog of Time, it would be good to get rid of that too.
3. Nothing. I didn't rule this out. You're looking at it.